cosmos/x/evm/store/cachekv/readonly_store.go

```diff
@@ -0,0 +1,46 @@
+// SPDX-License-Identifier: BUSL-1.1
+//
+// Copyright (C) 2023, Berachain Foundation. All rights reserved.
+// Use of this software is govered by the Business Source License included
+// in the LICENSE file of this repository and at www.mariadb.com/bsl11.
+//
+// ANY USE OF THE LICENSED WORK IN VIOLATION OF THIS LICENSE WILL AUTOMATICALLY
+// TERMINATE YOUR RIGHTS UNDER THIS LICENSE FOR THE CURRENT AND ALL OTHER
+// VERSIONS OF THE LICENSED WORK.
+//
+// THIS LICENSE DOES NOT GRANT YOU ANY RIGHT IN ANY TRADEMARK OR LOGO OF
+// LICENSOR OR ITS AFFILIATES (PROVIDED THAT YOU MAY USE A TRADEMARK OR LOGO OF
+// LICENSOR AS EXPRESSLY REQUIRED BY THIS LICENSE).
+//
+// TO THE EXTENT PERMITTED BY APPLICABLE LAW, THE LICENSED WORK IS PROVIDED ON
+// AN “AS IS” BASIS. LICENSOR HEREBY DISCLAIMS ALL WARRANTIES AND CONDITIONS,
+// EXPRESS OR IMPLIED, INCLUDING (WITHOUT LIMITATION) WARRANTIES OF
+// MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE, NON-INFRINGEMENT, AND
+// TITLE.
+
+package cachekv
+
+import (
+	storetypes "cosmossdk.io/store/types"
+
+	"pkg.berachain.dev/polaris/eth/core/vm"
+)
+
+var _ storetypes.CacheKVStore = (*ReadOnlyStore)(nil)
+
+// ReadOnlyStore is a wrapper around cachekv.Store that panics on any write operation.
+type ReadOnlyStore struct {
```

Comment:
wait , remind me why we can't use a saved err here?

Reply:
Geth catches the error as soon as a write op code is called. Similarly here we can catch it as soon as a write happens. The panic handler also catches other precompile panics as well